File: backend/services/forma_sync/package_builder.py
# -*- coding: utf-8 -*-
from __future__ import annotations

import json
import os
import tempfile
import zipfile
from dataclasses import dataclass
from typing import Any

from backend.services.forma_sync import sync_meta
from backend.services.forma_sync.entity_mappers import PackageMeta, rows_to_jsonl, sha256_file
from backend.services.forma_sync.entity_types import ENTITY_FILES, FormaSyncEntityType
from backend.services.forma_sync.export_changes import export_baseline_changes, export_pending_changes


@dataclass
class BuildPackageResult:
    zip_path: str
    sha256: str
    entities_summary: dict[str, int]
    row_count: int
    exported_refs: list[Any]


def _assemble_package(
    exported: dict,
    base_revision: int,
    *,
    baseline: bool = False,
) -> BuildPackageResult | None:
    if exported["row_count"] == 0:
        return None

    device_id = sync_meta.get_or_create_device_id()
    jsonl = exported["jsonl"]
    entities_summary = {entity: len(jsonl[entity]) for entity in ENTITY_FILES}

    tmp = tempfile.NamedTemporaryFile(suffix=".zip", delete=False)
    tmp_path = tmp.name
    tmp.close()

    from backend.services.forma_sync.entity_mappers import now_iso

    meta = PackageMeta(
        schema_version=1,
        device_id=device_id,
        source="desktop",
        created_at=now_iso(),
        base_revision=base_revision,
    )

    with zipfile.ZipFile(tmp_path, "w", compression=zipfile.ZIP_DEFLATED) as zf:
        zf.writestr("meta.json", json.dumps(meta.to_dict(), indent=2, ensure_ascii=False))
        for entity in ENTITY_FILES:
            content = rows_to_jsonl(jsonl[entity])
            if content:
                zf.writestr(f"changes/{entity}.jsonl", content)

    digest = sha256_file(tmp_path)
    return BuildPackageResult(
        zip_path=tmp_path,
        sha256=digest,
        entities_summary=entities_summary,
        row_count=exported["row_count"],
        exported_refs=exported["exported_refs"],
    )


def build_forma_sync_package(base_revision: int) -> BuildPackageResult | None:
    exported = export_pending_changes()
    return _assemble_package(exported, base_revision, baseline=False)


def build_forma_sync_baseline_package(base_revision: int = 0) -> BuildPackageResult | None:
    exported = export_baseline_changes()
    return _assemble_package(exported, base_revision, baseline=True)

File: mobile/src/sync/packageBuilder.ts
import RNFS from 'react-native-fs';

import {zip} from 'react-native-zip-archive';



import {initDB, nowIso} from '../database/index';

import {getOrCreateDeviceId} from './deviceId';

import {exportBaselineChanges} from './exportBaselineChanges';
import {exportPendingChanges} from './exportChanges';

import type {FormaSyncEntityType, FormaSyncJsonlRow} from './entityTypes';

import type {FormaSyncEntitiesSummary} from './manifest';

import {sha256HexFile} from './sha256File';



export type PackageMeta = {

  schema_version: 1;

  device_id: string;

  source: 'mobile';

  created_at: string;

  base_revision: number;

};



export type BuildPackageResult = {

  zipPath: string;

  sha256: string;

  entitiesSummary: FormaSyncEntitiesSummary;

  rowCount: number;

  exportedRefs: import('./changeTracker').ExportedEntityRef[];

};



function rowsToJsonl(rows: FormaSyncJsonlRow[]): string {

  if (rows.length === 0) {

    return '';

  }

  return rows.map(r => JSON.stringify(r)).join('\n') + '\n';

}



export async function buildFormaSyncPackage(
  baseRevision: number,
  options?: {baseline?: boolean},
): Promise<BuildPackageResult | null> {
  await initDB();

  const exported = options?.baseline
    ? await exportBaselineChanges()
    : await exportPendingChanges();

  if (exported.rowCount === 0) {
    return null;
  }



  const deviceId = await getOrCreateDeviceId();

  const entitiesSummary: FormaSyncEntitiesSummary = {

    food_entries: exported.jsonl.food_entries.length,

    body_metrics: exported.jsonl.body_metrics.length,

    strength_workouts: exported.jsonl.strength_workouts.length,

    stretching_log: exported.jsonl.stretching_log.length,

    bracelet_calories: exported.jsonl.bracelet_calories.length,

    hc_days: exported.jsonl.hc_days.length,

    cardio_workouts: exported.jsonl.cardio_workouts.length,

    food_products: exported.jsonl.food_products.length,

    strength_presets: exported.jsonl.strength_presets.length,

    user_preferences: exported.jsonl.user_preferences.length,

  };



  const dir = `${RNFS.CachesDirectoryPath}/forma-sync-build-${Date.now()}`;

  const changesDir = `${dir}/changes`;

  await RNFS.mkdir(dir);

  await RNFS.mkdir(changesDir);



  const meta: PackageMeta = {

    schema_version: 1,

    device_id: deviceId,

    source: 'mobile',

    created_at: nowIso(),

    base_revision: baseRevision,

  };

  await RNFS.writeFile(`${dir}/meta.json`, JSON.stringify(meta, null, 2), 'utf8');



  for (const [entity, rows] of Object.entries(exported.jsonl) as [FormaSyncEntityType, FormaSyncJsonlRow[]][]) {

    if (rows.length > 0) {

      await RNFS.writeFile(`${changesDir}/${entity}.jsonl`, rowsToJsonl(rows), 'utf8');

    }

  }



  const zipPath = `${RNFS.CachesDirectoryPath}/forma-sync-pkg-${Date.now()}.zip`;

  await zip(dir, zipPath);

  const sha256 = await sha256HexFile(zipPath);

  await RNFS.unlink(dir).catch(() => undefined);



  return {

    zipPath,

    sha256,

    entitiesSummary,

    rowCount: exported.rowCount,

    exportedRefs: exported.exportedRefs,

  };

}


